README.md
<p align="center">
  <img src="https://raw.githubusercontent.com/simplegoals/simplegoals/master/logo.png" alt="SimpleGoals: Beautiful and Easy-to-Use Achievements" width="40%">
</p>

<p align="center">Motivate your users with pleasant and easy-to-use achievements.</p>

<p align="center">
  <img src="https://raw.githubusercontent.com/simplegoals/simplegoals/master/demo.gif" alt="SimpleGoals: Beautiful and Easy-to-Use Achievements">
</p>

<p align="center"><b>Demo and a prettier version of this documentation is available at <a href="https://simplegoals.co">simplegoals.co</a></b></p>

**SimpleGoals** allows you to quickly and easily create achievements your users can accomplish — add intriguing game elements to your website.

And even more: **CloudStorage** allows you to keep users' achievements persistent across different browsers and devices. It only takes a couple of minutes to add new options to get everything ready.

- [How it works?](#how-does-it-work)
- [Installation](#installation)
  - [CDN (Easiest)](#cdn-easiest)
  - [Download files](#download-files)
  - [Webpack or Browserify (Advanced)](#webpack-or-browserify-advanced)
- [Usage](#usage)
  - [Initializing](#initializing)
  - [Unlock achievements](#unlock-achievements)
  - [Show all achievements](#show-all-achievements)
  - [Options](#options)
- [CloudStorage](#cloudstorage)
- [Development](#development)
- [License](#license)

# How does it work?

- Create a list of goals your users can achieve. 🏆
- Decide when each achievement is unlocked (e.g. on a button click, or when a special page was opened) ⏰
- Add a small JS snippet to your website. 👩‍💻
- Tada! 🎉

# Installation

There are several ways to get started with SimpleGoals. You can use:

### CDN (Easiest)
Just add this line to HTML code it before the closing `</body>` element.
```html
<script type="text/javascript" src="https://static.simplegoals.co/0.3.1/simplegoals-styles-included.min.js"></script>
```
And you're ready to use SimpleGoals!

If you're going to update styles of SimpleGoals, you might want to load JS and styles separately:
```html
// Before closing </head> element
<link rel="stylesheet" href="https://static.simplegoals.co/0.3.1/simplegoals.min.css">
// Before closing </body> element
<script type="text/javascript" src="https://static.simplegoals.co/0.3.1/simplegoals.min.js"></script>
```

### Download files
You can download SimpleGoals files from https://static.simplegoals.co/0.3.1/simplegoals.zip and add them to your project directory

Then load them in HTML:
```html
// Before closing </head> element
<link rel="stylesheet" href="simplegoals.min.css">
// Before closing </body> element
<script type="text/javascript" src="simplegoals.min.js"></script>
```

### Webpack or Browserify (Advanced)

Add package:
```shell
// With npm
npm install simplegoals --save

// With yarn
yarn add simplegoals
```

Load SimpleGoals in your JS code
```js
// Webpack
import 'simplegoals/dist/simplegoals.min.css'
import SimpleGoals from 'simplegoals';

// Browserify
require 'simplegoals/dist/simplegoals.min.css'
var SimpleGoals = require('simplegoals');
```

# Usage

Using SimpleGoals is fast and easy!

## Initializing

To use SimpleGoals you need to initialize it:
```js
// Define a list of all available goals (Example)
var goals = {
  explorer: {
    name: "Curious Explorer",
    description: "Scroll through half of the page"
  },
  secret: {
    name: "Lucky Digger",
    description: "Find a secret button"
  }
}

// Initialize SimpleGoals when a page is loaded
window.addEventListener('load', function() {
  SimpleGoals.init({
    goals: goals
  })
});
```
You can check out all available options in the [Options](#options) section

## Unlock achievements

There are two ways to unlock goals.
 - You can add `data-simplegoals-unlock="key_of_goal"` attribute to any button. When a user clicks this button, the goals will be unlocked and achievement will appear.
 ```html
 <a data-simplegoals-unlock="secret" href="#">I'm a secret button</a>
 ```
 - You can call `SimpleGoals.unlock("key_of_goal")` method at any time.
```js
SimpleGoals.unlock("explorer")
```
Example:
 ```js
 // Scroll through half of the page
 if (document.body.scrollTop > document.body.scrollHeight / 2) {
   SimpleGoals.unlock("explorer")
 }
 ```
 Each goal can be unlocked and shown only once.

 ## Show all achievements

There are also two ways to show all achievements.
 - You can add `data-simplegoals-overview="true"` attribute to any button. When user click this button, the list of all achievements will be shown. Example:
 ```html
 <a data-simplegoals-overview="true" href="#">Show all achievements</a>
 ```
 - You can call `SimpleGoals.showOverview()` method at any time.
 ```js
 SimpleGoals.showOverview()
 ```
 Example:
 ```js
 // Scroll through half of the page
 if (document.body.scrollTop > document.body.scrollHeight / 2) {
   SimpleGoals.showOverview()
 }
 ```

## Options


Option | Default | Description
-----|---------|------------
**`goals`** | `[]` |  An `Object` with all possible goals. Each key is a key of the goal and each value should be an `Object` with `name` and `description`
**`useCloudStorage`** | `false` | If `true`, **[CloudStorage](#CloudStorage)** will be used and goals for users will be saved in the cloud and will be persistent across all browser sessions and devices of the user. If `false`, `LocalStorage` will be used. All data will be saved in the user's browser and will be not synced. If a user clears local storage or use another browser, all achievements will be lost.
**`theme`** | see **[Theme Options](#theme-options)** | An `Object`, see **[Theme Options](#theme-options)**
**`onGoalUnlock`**  | `function(goalKey){}` | This is callback method that will be called, when a goal will be unlocked. It takes a goal's `key` as a param. You can use this callback method to add custom logging or some API calls.
**`freshStart`** | `false` | When `true` resets all goals every time a user refreshes or opens a new page. Works only when `useCloudStorage` is `false`.
**`timeout`** | `0` | This options sets the time(in milliseconds) between achievement appears and automatically disappears. If it is set to `0`, achievements will be not hidden automatically.
**`appId`** | `null` | Used only when **[CloudStorage](#CloudStorage)** is enabled. You can get your `appId` here: **https://api.simplegoals.co/projects/new**
**`user`** | `null` | Used only when **[CloudStorage](#CloudStorage)** is enabled. Used to determine user in CloudStorage. Object with `uid`, `email`, and `name`. Either `uid` or `email` should be present.

### Theme options
Option | Default | Description
-----|---------|------------
**`fontFamily`** | Browser default | Specifies the font for an elements
**`background`** | `#ffffff` | Background color
**`primary`** | `#38b2ac` | The color of buttons and links
**`title`** | `#000000` | The color of main labels
**`subtitle`** | `#718096` | The color of additional text
**`primaryHover`** | 7% darker than `primary` | The color of buttons and links in the hovered state, calculated automatically if not specified.
**`opposite`** | `#000000` or `#FFFFFF` | The color of text on buttons, calculated automatically if not specified.

Example:
```js
SimpleGoals.init({
  goals: goals,
  onGoalUnlock: function(goalKey){ console.log(goalKey) },
  theme: {
    fontFamily: 'Arial, Helvetica, sans-serif',
    background: '#ffffff',
    primary: '#38b2ac',
    title: '#000000',
    subtitle: '#718096'
  }
  timeout: 10000,
  useCloudStorage: true,
  appId: "f072c862-4f88-4be9-a86e-521e4a282e68",
  user: {
    uid: "magician#3123936",
    email: "harry@hogwarts.com",
    name: "Harry Potter"=
  }
})
```

# CloudStorage

CloudStorage allows you to keep users' achievements persistent across different browsers and devices. It only takes a couple of minutes to add new options to get everything ready.

There are two little steps you need to do:
 - Go to **https://api.simplegoals.co/projects/new** and get an **appId** that you can use. It's might be useful to get new **appId** for each of your environments: test, staging, production and so on.
 - Add new options to `SimpleGoals.init()` call:
 ```js
 SimpleGoals.init({
  goals: goals,
  useCloudStorage: true, // Required for CloudStorage
  appId: "yourAppId", // appId that you just got
  user: { // Information about User
    uid: "magician#3123936", // Optional, but either UID or email should be present
    email: "harry@hogwarts.com", // Optional, but either UID or email should be present
    name: "Harry Potter" // Optional
  }
})
 ```
UID is any unique identifier for a user.

Either UID or email should be present for user params.

*Note: It works only for logged in users*

*Note: When users first time converts from a guest to a logged in user, all their achievements are saved in CloudStorage*

# Development

First of all, thank you for making SimpleGoals better!

All source files are in `src` folder.

To install all dependecies run:
```shell
npm install
// or
yarn install
```

To run local environment run:
```shell
npm run playground
// or
yarn playground
```
Playground will be available at `http://localhost:8080`

To build production-ready bundle run:
```shell
npm run build
// or
yarn build
```

# License
This project is licensed under [MIT license](https://opensource.org/licenses/MIT).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.3.1] - 2019-11-25
### Fixed
- Polyfills were not generated

## [0.3.0] - 2019-11-03
### Added
- Added `theme` option

### Changed
- README text
- Added demo GIF
- Refactored JS code

## [0.2.0] - 2019-10-20
### Added
- Added `freshStart` option

## [0.1.3] - 2019-10-20
### Fixed
- Removed errors when `unlock` or `showOverview` methods are called before `init`

## [0.1.2] - 2019-10-19
### Changed
- Updated distribution
- Added new installation instructions to README
- Made options prettier in README

## [0.1.1] - 2019-10-17
### Added
- This CHANGELOG file
- Release script to package.json
### Fixed
- Imports for moduled systems

## [0.1.0] - 2019-10-17
### Added
- First working version of SimpleGoals

playground/simplegoals.js
(function webpackUniversalModuleDefinition(root, factory) {
	if(typeof exports === 'object' && typeof module === 'object')
		module.exports = factory();
	else if(typeof define === 'function' && define.amd)
		define("SimpleGoals", [], factory);
	else if(typeof exports === 'object')
		exports["SimpleGoals"] = factory();
	else
		root["SimpleGoals"] = factory();
})(window, function() {
return /******/ (function(modules) { // webpackBootstrap
/******/ 	// The module cache
/******/ 	var installedModules = {};
/******/
/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {
/******/
/******/ 		// Check if module is in cache
/******/ 		if(installedModules[moduleId]) {
/******/ 			return installedModules[moduleId].exports;
/******/ 		}
/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = installedModules[moduleId] = {
/******/ 			i: moduleId,
/******/ 			l: false,
/******/ 			exports: {}
/******/ 		};
/******/
/******/ 		// Execute the module function
/******/ 		modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);
/******/
/******/ 		// Flag the module as loaded
/******/ 		module.l = true;
/******/
/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}
/******/
/******/
/******/ 	// expose the modules object (__webpack_modules__)
/******/ 	__webpack_require__.m = modules;
/******/
/******/ 	// expose the module cache
/******/ 	__webpack_require__.c = installedModules;
/******/
/******/ 	// define getter function for harmony exports
/******/ 	__webpack_require__.d = function(exports, name, getter) {
/******/ 		if(!__webpack_require__.o(exports, name)) {
/******/ 			Object.defineProperty(exports, name, { enumerable: true, get: getter });
/******/ 		}
/******/ 	};
/******/
/******/ 	// define __esModule on exports
/******/ 	__webpack_require__.r = function(exports) {
/******/ 		if(typeof Symbol !== 'undefined' && Symbol.toStringTag) {
/******/ 			Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
/******/ 		}
/******/ 		Object.defineProperty(exports, '__esModule', { value: true });
/******/ 	};
/******/
/******/ 	// create a fake namespace object
/******/ 	// mode & 1: value is a module id, require it
/******/ 	// mode & 2: merge all properties of value into the ns
/******/ 	// mode & 4: return value when already ns object
/******/ 	// mode & 8|1: behave like require
/******/ 	__webpack_require__.t = function(value, mode) {
/******/ 		if(mode & 1) value = __webpack_require__(value);
/******/ 		if(mode & 8) return value;
/******/ 		if((mode & 4) && typeof value === 'object' && value && value.__esModule) return value;
/******/ 		var ns = Object.create(null);
/******/ 		__webpack_require__.r(ns);
/******/ 		Object.defineProperty(ns, 'default', { enumerable: true, value: value });
/******/ 		if(mode & 2 && typeof value != 'string') for(var key in value) __webpack_require__.d(ns, key, function(key) { return value[key]; }.bind(null, key));
/******/ 		return ns;
/******/ 	};
/******/
/******/ 	// getDefaultExport function for compatibility with non-harmony modules
/******/ 	__webpack_require__.n = function(module) {
/******/ 		var getter = module && module.__esModule ?
/******/ 			function getDefault() { return module['default']; } :
/******/ 			function getModuleExports() { return module; };
/******/ 		__webpack_require__.d(getter, 'a', getter);
/******/ 		return getter;
/******/ 	};
/******/
/******/ 	// Object.prototype.hasOwnProperty.call
/******/ 	__webpack_require__.o = function(object, property) { return Object.prototype.hasOwnProperty.call(object, property); };
/******/
/******/ 	// __webpack_public_path__
/******/ 	__webpack_require__.p = "";
/******/
/******/
/******/ 	// Load entry module and return exports
/******/ 	return __webpack_require__(__webpack_require__.s = "./src/index.js");
/******/ })
/************************************************************************/
/******/ ({

/***/ "./src/index.css":
/*!***********************!*\
  !*** ./src/index.css ***!
  \***********************/
/*! no static exports found */
/***/ (function(module, exports, __webpack_require__) {

// extracted by mini-css-extract-plugin

/***/ }),

/***/ "./src/index.js":
/*!**********************!*\
  !*** ./src/index.js ***!
  \**********************/
/*! exports provided: init, unlock, showOverview, default */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

"use strict";
__webpack_require__.r(__webpack_exports__);
/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, "init", function() { return init; });
/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, "unlock", function() { return unlock; });
/* harmony import */ var _modules_local_goals__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! ./modules/local-goals */ "./src/modules/local-goals.js");
/* harmony import */ var _modules_cloud_storage__WEBPACK_IMPORTED_MODULE_1__ = __webpack_require__(/*! ./modules/cloud-storage */ "./src/modules/cloud-storage.js");
/* harmony import */ var _modules_overview__WEBPACK_IMPORTED_MODULE_2__ = __webpack_require__(/*! ./modules/overview */ "./src/modules/overview.js");
/* harmony import */ var _modules_achievements__WEBPACK_IMPORTED_MODULE_3__ = __webpack_require__(/*! ./modules/achievements */ "./src/modules/achievements.js");
/* harmony import */ var _modules_buttons__WEBPACK_IMPORTED_MODULE_4__ = __webpack_require__(/*! ./modules/buttons */ "./src/modules/buttons.js");
/* harmony reexport (safe) */ __webpack_require__.d(__webpack_exports__, "showOverview", function() { return _modules_overview__WEBPACK_IMPORTED_MODULE_2__["showOverview"]; });

__webpack_require__ (/*! ./index.css */ "./src/index.css")







let options = {
  timeout: 0,
  onGoalUnlock: (goalKey) => {},
  freshStart: false,
  useCloudStorage: false,
  appId: null,
  user: {}
}
let goals = null

const init = async config => {
  Object.assign(options, config)
  goals = Object(_modules_local_goals__WEBPACK_IMPORTED_MODULE_0__["prepareGoals"])(options)
  await Object(_modules_cloud_storage__WEBPACK_IMPORTED_MODULE_1__["loadFromCloud"])(goals, options)
  Object(_modules_overview__WEBPACK_IMPORTED_MODULE_2__["initOverview"])(goals, options)
  Object(_modules_buttons__WEBPACK_IMPORTED_MODULE_4__["initButtons"])(goals, options)
}

const unlock = async (name) => {
  Object(_modules_achievements__WEBPACK_IMPORTED_MODULE_3__["unlockGoal"])(name, goals, options)
}



/* harmony default export */ __webpack_exports__["default"] = ({init, unlock, showOverview: _modules_overview__WEBPACK_IMPORTED_MODULE_2__["showOverview"]});


/***/ }),

/***/ "./src/modules/achievements.js":
/*!*************************************!*\
  !*** ./src/modules/achievements.js ***!
  \*************************************/
/*! exports provided: unlockGoal */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

"use strict";
__webpack_require__.r(__webpack_exports__);
/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, "unlockGoal", function() { return unlockGoal; });
/* harmony import */ var _support__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! ./support */ "./src/modules/support.js");
/* harmony import */ var _local_goals__WEBPACK_IMPORTED_MODULE_1__ = __webpack_require__(/*! ./local-goals */ "./src/modules/local-goals.js");
/* harmony import */ var _cloud_storage__WEBPACK_IMPORTED_MODULE_2__ = __webpack_require__(/*! ./cloud-storage */ "./src/modules/cloud-storage.js");
/* harmony import */ var _overview__WEBPACK_IMPORTED_MODULE_3__ = __webpack_require__(/*! ./overview */ "./src/modules/overview.js");
/* harmony import */ var _styles__WEBPACK_IMPORTED_MODULE_4__ = __webpack_require__(/*! ./styles */ "./src/modules/styles.js");






const createAchievementHtml = (key, goal, button, options) => {
  const achievementHtmlString = `
  <div class="simplegoals-achievement" id="simplegoals-achievement-${key}">
    <div class="simplegoals-achievement__icon">
      <img src="https://static.simplegoals.co/cup.svg" />
    </div>
    <div class="simplegoals-achievement__body">
      <p class="simplegoals-achievement__notification">
        New achievement unlocked
      </p>
      <p class="simplegoals-achievement__name" id="simplegoals-achievement__name">${goal.name}</p>
    </div>
    <div class="simplegoals-achievement__button" id="simplegoals-achievement__button-${key}" role="button">${button}</div>
    <button class="simplegoals-achievement__close-button" id="simplegoals-achievement__close-button-${key}">
      <svg viewBox="0 0 500 500" fill="none" xmlns="http://www.w3.org/2000/svg">
        <g>
          <path d="M50 50L450 450" stroke-width="100" stroke-linecap="round"/>
          <path d="M50 450L450 50" stroke-width="100" stroke-linecap="round"/>
        </g>
      </svg>
    </button>
  </div>
  `
  const node = new DOMParser().parseFromString(achievementHtmlString , 'text/html').body.firstChild
  Object(_styles__WEBPACK_IMPORTED_MODULE_4__["applyStyles"])(node, options)
  document.body.appendChild(node)
  return node
}

const recalculateAchievementsTop = () => {
  const nodes = document.getElementsByClassName('simplegoals-achievement')
  for(let i=0; i < nodes.length; i++) {
    nodes[i].style.top = `${1 + i*7}em`
  }
}

const showAchievement = (name, goal, options) => {
  const key = Object(_support__WEBPACK_IMPORTED_MODULE_0__["hashCode"])(name)
  const achievement = createAchievementHtml(key, goal, 'Learn more', options)
  recalculateAchievementsTop()
  achievement.classList.add('simplegoals-achievement--opened')
  initAchievement(key, achievement)
  if (options.timeout) {
    achievementTimeout = setTimeout(() => hideAchievement(achievement), options.timeout)
  }
}

const initAchievement = (key, achievement) => {
  achievement.addEventListener('click', event => achievement.classList.toggle('simplegoals-achievement--clicked'))
  const trigger = document.getElementById(`simplegoals-achievement__close-button-${key}`)
  trigger.addEventListener('click', event => hideAchievement(achievement))
  const overviewTrigger = document.getElementById(`simplegoals-achievement__button-${key}`)
  overviewTrigger.addEventListener('click', event => {
    hideAchievement(achievement)
    setTimeout(_overview__WEBPACK_IMPORTED_MODULE_3__["showOverview"], 200)
  })
}

const hideAchievement = (achievement) => {
  achievement.classList.remove('simplegoals-achievement--opened')
  achievement.classList.add('simplegoals-achievement--closed')
  setTimeout(() => {
    achievement.parentNode.removeChild(achievement)
    recalculateAchievementsTop()
  }, 2000)
}

const unlockGoal = async (name, goals, options) => {
  if(!goals || !goals[name]) { return }
  if(goals[name].unlocked){
    return
  }
  goals[name].unlocked = true
  options.onGoalUnlock(name)
  Object(_local_goals__WEBPACK_IMPORTED_MODULE_1__["saveLocalGoal"])(name, options)
  await Object(_cloud_storage__WEBPACK_IMPORTED_MODULE_2__["saveCloudGoal"])(name, options)
  showAchievement(name, goals[name], options)
  Object(_overview__WEBPACK_IMPORTED_MODULE_3__["rerenderOverviewGoals"])(goals)
}

/***/ }),

/***/ "./src/modules/buttons.js":
/*!********************************!*\
  !*** ./src/modules/buttons.js ***!
  \********************************/
/*! exports provided: initButtons */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

"use strict";
__webpack_require__.r(__webpack_exports__);
/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, "initButtons", function() { return initButtons; });
/* harmony import */ var _overview__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! ./overview */ "./src/modules/overview.js");
/* harmony import */ var _achievements__WEBPACK_IMPORTED_MODULE_1__ = __webpack_require__(/*! ./achievements */ "./src/modules/achievements.js");



const initButtons = (goals, options) => {
  document.addEventListener('click', function (event) {
    if (event.target.dataset.simplegoalsUnlock) {
      event.preventDefault()
      Object(_achievements__WEBPACK_IMPORTED_MODULE_1__["unlockGoal"])(event.target.dataset.simplegoalsUnlock, goals, options)
    }
    if (event.target.dataset.simplegoalsOverview) {
      event.preventDefault()
      Object(_overview__WEBPACK_IMPORTED_MODULE_0__["showOverview"])();
    }
  }, false);
}

/***/ }),

/***/ "./src/modules/cloud-storage.js":
/*!**************************************!*\
  !*** ./src/modules/cloud-storage.js ***!
  \**************************************/
/*! exports provided: saveCloudGoal, loadFromCloud */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

"use strict";
__webpack_require__.r(__webpack_exports__);
/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, "saveCloudGoal", function() { return saveCloudGoal; });
/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, "loadFromCloud", function() { return loadFromCloud; });
/* harmony import */ var _support__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! ./support */ "./src/modules/support.js");


const saveCloudGoal = async (name, options) => {
  if (options.useCloudStorage) {
    let data = {
      name: name
    }
    addAuthData(data, options)
    await Object(_support__WEBPACK_IMPORTED_MODULE_0__["postRequest"])("/goals/unlock", data)
  } else {
    return false
  }
}

const addAuthData = (hash, options) => {
  hash.appId = options.appId
  hash.user = options.user
}

const loadFromCloud = async (goals, options) => {
  if (options.useCloudStorage) {
    let data = {
      goals: Object.keys(goals).map((key) => {
        return {
          name: key,
          description: goals[key].description
        }
      }),
      unlockedGoals: Object.keys(goals).filter((key) => {
        goals[key].unlocked
      })
    }
    addAuthData(data, options)
    const response = await Object(_support__WEBPACK_IMPORTED_MODULE_0__["postRequest"])("/sessions", data)
    if (response) {
      updateGoalsFromCloud(response, goals)
    }
  } else {
    return false
  }
}

const updateGoalsFromCloud = (response, goals) => {
  for (const key of Object.keys(goals)) {
    goals[key].unlocked = response.unlockedGoals.includes(key)
  }
}

/***/ }),

/***/ "./src/modules/local-goals.js":
/*!************************************!*\
  !*** ./src/modules/local-goals.js ***!
  \************************************/
/*! exports provided: saveLocalGoal, prepareGoals */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

"use strict";
__webpack_require__.r(__webpack_exports__);
/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, "saveLocalGoal", function() { return saveLocalGoal; });
/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, "prepareGoals", function() { return prepareGoals; });
const saveLocalGoal = (name, options) => {
  if (typeof(Storage) !== "undefined" && !options.freshStart) {
    let localGoals = getLocalGoals(options)
    localGoals.push(name)
    localStorage.setItem('simplegoals-storage', JSON.stringify(localGoals))
  }
}

const getLocalGoals = (options) => {
  if (typeof(Storage) !== "undefined" && !options.freshStart) {
    return JSON.parse(localStorage.getItem('simplegoals-storage')) || []
  } else {
    return []
  }
}

const prepareGoals = (options) => {
  let goals = options.goals
  const localGoals = getLocalGoals(options)
  for (const key of Object.keys(goals)) {
    goals[key].unlocked = localGoals.includes(key)
  }
  return goals
}

/***/ }),

/***/ "./src/modules/overview.js":
/*!*********************************!*\
  !*** ./src/modules/overview.js ***!
  \*********************************/
/*! exports provided: initOverview, rerenderOverviewGoals, showOverview */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

"use strict";
__webpack_require__.r(__webpack_exports__);
/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, "initOverview", function() { return initOverview; });
/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, "rerenderOverviewGoals", function() { return rerenderOverviewGoals; });
/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, "showOverview", function() { return showOverview; });
/* harmony import */ var _styles__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! ./styles */ "./src/modules/styles.js");


let overview = null

const initOverview = (goals, options) => {
  createOverviewHtml(goals, options)
  overview = document.getElementById('simplegoals-overview')
  const trigger = document.getElementById('simplegoals-overview__close-button')
  trigger.addEventListener('click', event => hideOverview())
}

const createOverviewHtml = (goals, options) => {
  const overviewHtmlString = `
  <div class="simplegoals-overview" id="simplegoals-overview">
    <div class="simplegoals-overview__wrapper">
      <div class="simplegoals-overview__body">
        <h1 class="simplegoals-overview__title">Achievements</h1>
        <p class="simplegoals-overview__subtitle">Complete tasks and unlock achievements</p>
        <div class="simplegoals-overview__goals" id="simplegoals-overview__goals">
        ` + overviewGoalsHtml(goals) + `
        </div>
        <button class="simplegoals-overview__close-button" id="simplegoals-overview__close-button">
          Close
        </button>
        <div class="simplegoals-overview__attribution">
          Powered by <a class="simplegoals-overview__attribution-link" href="https://simplegoals.co" target="_blank">SimpleGoals</a>
        </div>
      </div>
    </div>
  </div>
  `
  const node = new DOMParser().parseFromString(overviewHtmlString , 'text/html').body.firstChild
  Object(_styles__WEBPACK_IMPORTED_MODULE_0__["applyStyles"])(node, options)
  document.body.appendChild(node)
}

const overviewGoalsHtml = (goals) => {
  let result = "";
  const sortedKeys = Object.keys(goals).sort((a, b) => goals[b].unlocked - goals[a].unlocked)
  for (const key of sortedKeys) {
    result += `
    <div class="simplegoals-overview-goal ${goals[key].unlocked ? "" : "simplegoals-overview-goal--locked"}">
      <div class="simplegoals-overview-goal__icon">
        <img src="https://static.simplegoals.co/cup.svg" />
      </div>
      <div class="simplegoals-overview-goal__body">
        <h3 class="simplegoals-overview-goal__name">${goals[key].name}</h3>
        <p class="simplegoals-overview-goal__description">
          ${goals[key].description}
        </p>
      </div>
    </div>
    `
  }
  return result
}

const rerenderOverviewGoals = (goals) => {
  document.getElementById('simplegoals-overview__goals').innerHTML = overviewGoalsHtml(goals)
}

const showOverview = () => {
  if(!overview) { return }
  overview.scrollTo(0, 0)
  overview.classList.add('simplegoals-overview--opened')
}

const hideOverview = () => {
  overview.classList.remove('simplegoals-overview--opened')
}

/***/ }),

/***/ "./src/modules/styles.js":
/*!*******************************!*\
  !*** ./src/modules/styles.js ***!
  \*******************************/
/*! exports provided: applyStyles */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

"use strict";
__webpack_require__.r(__webpack_exports__);
/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, "applyStyles", function() { return applyStyles; });
const defaultStyleOptions = {
  fontFamily: 'BlinkMacSystemFont, -apple-system, "Segoe UI", "Roboto", "Oxygen", "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans", "Helvetica Neue", "Helvetica", "Arial", sans-serif',
  background: '#ffffff',
  title: '#000000',
  subtitle: '#718096',
  primary: '#38b2ac',
  primaryHover: '#319795',
  opposite: '#ffffff'
}

const getStyles = (options = {}) => {
  let styles = defaultStyleOptions;
  ['fontFamily', 'background', 'title', 'subtitle'].forEach(key => {
    if (options[key]) {
      styles[key] = options[key]
    }
  })
  if (options.primary) {
    styles.primary = options.primary
    styles.primaryHover = getHover(options.primary)
    styles.opposite = getOpposite(options.primary)
  }
  ['primaryHover', 'opposite'].forEach(key => {
    if (options[key]) {
      styles[key] = options[key]
    }
  })
  return styles
}

const getHover = (color) => {
  const percent = 7

  let R = parseInt(color.substring(1,3),16);
  let G = parseInt(color.substring(3,5),16);
  let B = parseInt(color.substring(5,7),16);

  R = parseInt(R * (100 - percent) / 100);
  G = parseInt(G * (100 - percent) / 100);
  B = parseInt(B * (100 - percent) / 100);

  R = (R<255)?R:255;  
  G = (G<255)?G:255;  
  B = (B<255)?B:255;  

  const RR = ((R.toString(16).length==1)?"0"+R.toString(16):R.toString(16));
  const GG = ((G.toString(16).length==1)?"0"+G.toString(16):G.toString(16));
  const BB = ((B.toString(16).length==1)?"0"+B.toString(16):B.toString(16));

  return "#"+RR+GG+BB;
}

const getOpposite = (hex) => {
  if (hex.indexOf('#') === 0) {
    hex = hex.slice(1);
  }
  // convert 3-digit hex to 6-digits.
  if (hex.length === 3) {
      hex = hex[0] + hex[0] + hex[1] + hex[1] + hex[2] + hex[2];
  }
  if (hex.length !== 6) {
      throw new Error('Invalid HEX color.');
  }
  var r = parseInt(hex.slice(0, 2), 16),
      g = parseInt(hex.slice(2, 4), 16),
      b = parseInt(hex.slice(4, 6), 16);
  // http://stackoverflow.com/a/3943023/112731
  return (r * 0.299 + g * 0.587 + b * 0.114) > 186
      ? '#000000'
      : '#FFFFFF';
}

const applyStyles = (element, options) => {
  let styles = getStyles(options.theme)
  element.style.setProperty('--simplegoals-font-family', styles.fontFamily);
  element.style.setProperty('--simplegoals-color-background', styles.background);
  element.style.setProperty('--simplegoals-color-title', styles.title);
  element.style.setProperty('--simplegoals-color-subtitle', styles.subtitle);
  element.style.setProperty('--simplegoals-color-primary', styles.primary);
  element.style.setProperty('--simplegoals-color-primary-hover', styles.primaryHover);
  element.style.setProperty('--simplegoals-color-opposite', styles.opposite);
}

/***/ }),

/***/ "./src/modules/support.js":
/*!********************************!*\
  !*** ./src/modules/support.js ***!
  \********************************/
/*! exports provided: postRequest, hashCode */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

"use strict";
__webpack_require__.r(__webpack_exports__);
/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, "postRequest", function() { return postRequest; });
/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, "hashCode", function() { return hashCode; });
const apiUrl = "https://api.simplegoals.co"

const postRequest = async (endpoint, data) => {
  const response = await fetch(`${apiUrl}${endpoint}`, {
    method: 'POST',
    body: JSON.stringify(data),
    headers: {
      'Content-Type': 'application/json'
    }
  })
  if(response.ok) {
    return await response.json()
  } else {
    return false
  }
}

const hashCode = (source) => {
  let hash = 0;
  if (source.length == 0) { return hash; }
  for (let i = 0; i < source.length; i++) {
    let char = source.charCodeAt(i);
    hash = ((hash<<5)-hash)+char;
    hash = hash & hash; // Convert to 32bit integer
  }
  return hash;
}

/***/ })

/******/ });
});
//# sourceMappingURL=simplegoals.js.map